http-common/src/request.rs

```diff
@@ -98,73 +138,94 @@ where
     method: http::Method,
     uri: TUri,
     headers: Option<&[(&str, &str)]>,
+    max_retries: u32,
     body: Option<&TRequest>,
 ) -> std::io::Result<(hyper::StatusCode, hyper::HeaderMap, hyper::body::Bytes)>
 where
     hyper::Uri: TryFrom<TUri>,
     <hyper::Uri as TryFrom<TUri>>::Error: Into<http::Error>,
     TRequest: serde::Serialize,
+    TUri: Clone,
 {
-    let mut req = hyper::Request::builder().method(method).uri(uri);
-
-    if let Some(headers) = headers {
-        if let Some(headers_map) = req.headers_mut() {
-            for (key, value) in headers {
-                headers_map.insert(
-                    headers::HeaderName::from_str(key)
-                        .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))?,
-                    headers::HeaderValue::from_str(value)
-                        .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))?,
+    let mut retry_num = 0;
+    loop {
+        let mut req = hyper::Request::builder()
+            .method(method.clone())
+            .uri(uri.clone());
+
+        if let Some(headers) = headers {
+            if let Some(headers_map) = req.headers_mut() {
+                for (key, value) in headers {
+                    headers_map.insert(
+                        headers::HeaderName::from_str(key)
+                            .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))?,
+                        headers::HeaderValue::from_str(value)
+                            .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))?,
+                    );
+                }
+            }
+        }
+
+        // `req` is consumed by both branches, so this cannot be replaced with `Option::map_or_else`
+        //
+        // Ref: https://github.com/rust-lang/rust-clippy/issues/5822
+        #[allow(clippy::option_if_let_else)]
+        let req = if let Some(body) = body {
+            let body = serde_json::to_vec(body)
+                .expect("serializing request body to JSON cannot fail")
+                .into();
+
+            req.header(hyper::header::CONTENT_TYPE, "application/json")
+                .body(body)
+        } else {
+            req.body(Default::default())
+        };
+        let req = req.expect("cannot fail to create hyper request");
+
+        let response = client.request(req).await;
+        match response {
+            Ok(response) => {
+                let (
+                    http::response::Parts {
+                        status, headers, ..
+                    },
+                    body,
+                ) = response.into_parts();
+
+                let body = hyper::body::to_bytes(body)
+                    .await
+                    .map_err(|err| std::io::Error::new(std::io::ErrorKind::Other, err))?;
+
+                return Ok((status, headers, body));
+            }
+            Err(err) if err.is_closed() && retry_num < max_retries => {
+                log::warn!(
```

Comment:
Will this flood our logs? Let's check.. if it is not very common, then it is okay, else we can make it debug..

Reply:
It should be very uncommon.